### ISSUE TITLE
Q4 2023 priorities

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#4070]
In the fourth quarter of 2023, our priorities are:

* Implementing [C++23 features](https://github.com/microsoft/STL/projects/8) from the Working Draft
  + We have an active multi-collaborator feature branch for #2912
* Keeping PR reviews flowing, so contributions are reviewed promptly
* Improving STL ASan test coverage

If you have questions about our plans for this quarter, feel free to ask them here. (Note that compiler, IDE, etc. priorities are off-topic.)

There is [no news](https://github.com/microsoft/STL/issues/3628#issuecomment-1579449191) about open-sourcing VCRuntime/VCStartup.

**Comment by h-vetinari** [comment]
> compiler priorities are off-topic.

I have a question that's not about priorities, as much as curiosity about a strange divergence I noted on the [C++ support](https://en.cppreference.com/w/cpp/compiler_support) comparison from cppreference (already for a couple of quarters, but now it got to the point where I thought I'd ask).

According to the status page[^1], while the STL has implemented the vast majority of C++23 features already, the MSVC compiler has implemented barely any!?! Not sure if all those features are incubating somewhere in secret, but it struck me that the STL looks like it might soon be "done" with C++23, while the compiler looks _very_ far from it.

To bring the question back around to the STL - will work on C++26 be started after C++23 is done in the STL, even if assistance from the compiler on (C++26-specific) issues that arise is very far away?

[^1]: same story for the [status page](https://learn.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=msvc-170#c-compiler-features) maintained by Microsoft

**Comment by fsb4000** [comment]
@h-vetinari 

![изображение](https://github.com/microsoft/STL/assets/4289847/74968914-402c-448d-ab38-c81399e453d2)

**Comment by StephanTLavavej** [comment]
> will work on C++26 be started after C++23 is done in the STL

Maybe. After finishing all C++23 library-only features, we might start work on C++26 library-only features, we might focus on bugfixing, or we might start vNext (no idea if this will happen).

**Comment by jovibor** [comment]
When C++23 work is done, we'll have exactly **five** modes in VS: **C++14**, **C++17**, **C++20**, **C++23**, **C++latest**, which is quite a lot.

Is there any plans to drop at least **C++14** support, which is 10 years old, in STL?

**Comment by StephanTLavavej** [comment]
No. Won't happen for the v19 binary-compatible release series, ever.

Might happen for vNext if we're lucky. If and when vNext happens, I'll strongly recommend that we start with the current Standard as a baseline, given how a rebuild-the-world migration will be necessary to join the vNext world anyways.

**Comment by h-vetinari** [comment]
> If and when vNext happens, I'll strongly recommend that we start with the current Standard as a baseline, given how a rebuild-the-world migration will be necessary to join the vNext world anyways.

With my conda-forge hat on, that doesn't sound like a great idea, because it means that if we want to rebuild our ecosystem (thousands of packages) for vNext, we don't just have to do the rebuild, but _also_ ensure that all upstream projects (not under our control) are source-compatible with the newest C++ standard. You can bet your hat that a lot of them won't be, meaning a vNext migration would take many orders of magnitude longer than "just" rebuilding everything for a new ABI.

I'd be surprised if vcpkg wouldn't have the same issue too.

**Comment by jovibor** [comment]
> ensure that all upstream projects (not under our control) are source-compatible with the newest C++ standard.

If you somehow uncertain or in doubt, you just stay with the old toolset.
The world can't and should not wait for everyone to adopt to new standards, it would take eternity.

**Comment by h-vetinari** [comment]
> If you somehow uncertain or in doubt, you just stay with the old toolset.

My response got a bit long, and I don't want to derail the thread too much, so it's behind a fold-out.

<details>

The old toolset has the old ABI (right?), so it doesn't mix. That creates a deadend when you're in a large graph of packages where you first need all your dependencies built for the right ABI first in order to then build a given package (that build graph is essentially the situation of all package managers and distributions).

I'm 100% in favour of vNext, I'm just saying that if it only supports C++23 (or C++26 or ...) and up, it's going to be nigh-impossible to build projects with non-trivial third-party dependencies (until they all become compatible with said standard version), because all those have to be recompiled first. The only escape hatch I could see would be verifying that a given dependency is not affected by any of the ABI-differences in vNext (so you can link to an artefact compiled with vc14), but that's not feasible at scale, aside from being completely unsupported.

Taking a medium-sized project that's in widespread use (not just in C++ land), [arrow](https://github.com/apache/arrow/blob/main/docs/source/developers/cpp/building.rst), you'll need to have [abseil](https://github.com/abseil/abseil-cpp/), [aws-sdk-cpp](https://github.com/aws/aws-sdk-cpp), [boost](https://github.com/boostorg/boost), [glog](https://github.com/google/glog), [gflags](https://github.com/gflags/gflags), [google-cloud-cpp](https://github.com/googleapis/google-cloud-cpp), [grpc](https://github.com/grpc/grpc), [LLVM](https://github.com/llvm/llvm-project/tree/main/llvm), [protobuf](https://github.com/protocolbuffers/protobuf/), [orc](https://github.com/apache/orc), [re2](https://github.com/google/re2), [snappy](https://github.com/google/snappy), [thrift](https://github.com/apache/thrift) and all _their_ dependencies built for vNext before you can unblock all the packages depending on arrow (and its various bindings).

Basically, if vNext only supports really new C++ standards, adoption speed is going to be glacial. It's still possible that it's preferable to do it like that - I don't claim to know what the right trade-off is. But as a distributor, it would be _massively_ easier to roll out (and help spur adoption) if vNext supported at least C++17 and up. With C++20 as impactful & disruptive as it has been, few widely-used projects (especially those concerned about cross-platform and cross-compiler compatibility) have really started moving to C++20 in earnest.

</details>

**Comment by jovibor** [comment]
> But as a distributor, it would be massively easier to roll out (and help spur adoption) if vNext supported at least C++17 and up. 
<details><summary>Expand</summary>

You want **vNext** with the new ABI but old standard (C++17)? But what's the point? You'll need to recompile the world anyway for the new ABI. But why recompile it with the old standard, what's the logic behind this?

C++ standards are always backward compatible, except for some tiny moments like deprecating rare stuff. If you really need **C++17** (or 14) you just stay with the old toolset.

BTW, supporting many standards at once is very expensive, and STL maintainers are not an unlimited resource.
</details>

**Comment by h-vetinari** [comment]
I feel bad about how OT this is becoming, so this is my last entry in the discussion with @jovibor. 😅 

<details>

> You'll need to recompile the world anyway for the new ABI. But why recompile it with the old standard

Because the world is big and interconnected and not everyone is compatible with the newest standard. Requiring vNext to use the latest C++ standard would greatly slow adoption of vNext (and bug reports for, and require longer support for vNow, etc.).

It creates a coupling between the compiler, stdlib and the C++ standard that makes it a much bigger "cliff" to upgrade. A given project cannot separately test the new compiler, or just the newest standard (if vNext starts off with C++26) - you have to upgrade (& validate) both at the same time, plus ensure that all your dependencies survive that. It's a huge (t)ask IMO, so it'd be slooooow (which also isn't in the interest of Visual Studio as a product, IMO).

> C++ standards are always backward compatible

You'd be surprised how much things break between standards once you reach a scale of hundreds of projects and millions of lines of code. I'm not even talking about explicit removals like `std::{unary,binary}_function`, or bugs in the compiler/stdlib, but changes in handling of `const(expr)`, SFINAE etc. And don't get me started about the ABI-impacts of the standard version on any consumer of abseil.

> supporting many standards at once is very expensive, and STL maintainers are not an unlimited resource.

Fully aware of that, which is why I said I don't claim to know where the trade-off is. But it's not like keeping older standard modes around means everything needs to be rewritten for vNext. A single PR here could break the ABI (e.g. less than 20 lines for https://github.com/microsoft/STL/pull/1244); keeping old standard modes would mostly mean not deleting existing code and keeping it tested, which certainly has a cost, but is more like "keeping the lights on" rather than "rebuild Rome".

> If you really need **C++17** (or 14) you just stay with the old toolset.

I don't think you understand the constraints of large-scale package management (either binary distribution or something like vcpkg). Unless everything is built for a consistent ABI (with a consistent - or at least compatible - toolset, like vc140-vc143), then things blow up in spectacular fashion - segfaults, crashes, data corruption, etc. There's no alternative to building for a consistent ABI and building _everything_ includes projects that are not at the forefront of current C++ standards (but are often dependencies of other packages, so you cannot just skip them).

</details>

**Comment by StephanTLavavej** [comment]
> A given project cannot separately test the new compiler, or just the newest standard (if vNext starts off with C++26) - you have to upgrade (& validate) both at the same time, plus ensure that all your dependencies survive that.

I don't believe that this is correct. The v19 binary-compatible release series will support up to C++23 (`/std:c++23` doesn't exist yet but it will someday), so projects can upgrade all the way to C++23 while remaining binary-compatible. Even if vNext increases its minimum required Standard version significantly, we would surely overlap with v19's supported Standard (having v19 max out at C++23 and vNext start at C++26 would make no sense and achieve very little).

Therefore projects would be able to upgrade their Standard version (making source changes as necessary), and then switch to vNext (making different source changes, possibly), in separate steps.

Projects should be upgrading the Standard version that they can tolerate right *now*, in fact.

**Comment by h-vetinari** [comment]
> having v19 max out at C++23 and vNext start at C++26 would make no sense and achieve very little

Having common ground with C++23 is already a clear positive that allows a less steep upgrade path. 👍 Though that common ground could (and likely should) still be larger.

> Projects should be upgrading the Standard version that they can tolerate right _now_, in fact.

That's a nice idea in theory. [more below]

<details>

In practice, no toolchain has complete C++23 support, and while C++20 is pretty complete across different tools (aside from broader modules support obviously), projects are hesitant to upgrade the C++ standard too aggressively. It opens them up to compiler bugs resp. breaking changes (especially if the implementation in the compiler isn't complete yet), it cuts off large chunks of their user base on older distros, and so on.

Those are reasons why changing the default gets postponed (completely aside from the regular feature/maintenance demands), but it's similar on the testing side, because it's common for projects to just default to a C++ version in their builds and not test anything else (and we've all learned that what's not tested ends up broken).

As a concrete example, boost - _the_ C++ posterchild library - still had breakage with >=C++17 and contemporary clang [until](https://github.com/boostorg/functional/commit/6a573e4b8333ee63ee62ce95558c3667348db233) the very latest release (1.83)[^1] 2 months ago. Abseil removed C++11 support only a year ago, and is [planning](https://github.com/google/oss-policies-info/blob/main/foundational-cxx-support-matrix.md) to remove C++14 end of 2024. And so on.

This is not to say that nothing can ever be dropped (far from it!). Still, the stance of dropping all but one current standard modes just strikes me as very far removed from the actual situation in the broader ecosystem (obviously my perception is not objective, but as a distributor I see the settings/compatibility of ~hundreds of projects). From my POV, for well-maintained projects in the FOSS world, C++17 is nowadays becoming the default, but I know very few projects that default to (much less require) C++20, and the testing penetration of that is even worse.

The situation would probably be different if C++20 hadn't been such a huge release, but that's what it is.

</details>

[^1]: not the first time this happens either: https://github.com/boostorg/config/commit/c35885ced42096c3d8c8f0cda930d533d94acae4

**Comment by rboxman** [comment]
I've wanted v.next to happen for years now. Unfortunately, I'm forced to echo the above considerations as well.

<details>
If v.next requires C++26 as a base then that would make it DOA in my circle for many years to come. I can understand wanting to baseline the c++ version at the same time as breaking ABI but that simply means that v.next has arrived much too late; it should have happened sooner.


It will be a decade until the folks in the VFX industry will be able to use C++26 or similar. See https://vfxplatform.com/ and notice that C++20 isn't even listed yet. Projects literally cannot start using/shipping/supporting newer standards already. Hopefully 20 will be planned for 2025... unless some poor soul wants to convince Autodesk, Pixar, and SideFX to skip 20 and move directly to C++23 in 2025... But too bad MSVC language support for C++23 has been delayed to some indeterminant date in the future due to business problems and which will no doubt affect such a decision. The decision for what to put into 2025 will happen early spring next year. Mistake after compounding mistake here.


Windows is already a second class citizen in that circle and this would further cement that. The ripple effect to downstream projects is massive. There's a lot of highly visible projects that depend on those libraries and which need to support stable C++ SDK APIs. Highly visible projects with early in career folks/volunteers who will continue to see Windows and MSVC in an extremely poor light. If you want v.next to help strengthen the Windows ecosystem and get new blood involved baselining at C++26 is not the way to win them over.

If you've followed the VFX platform for a few years you'll know how long it took for them to switch away from GCC's older ABI, but instrumental in that effort was the ability to use a large set of language standards along the way. I'd also vote for C++23 being the lower floor rather than 26; but 20 if possible just in case.
</details>

**Comment by StephanTLavavej** [comment]
> If v.next requires C++26 as a base

To clarify my previous comment: if vNext happens any time this decade, there's no way we would make it require C++26.

No decisions have been made yet - this is just me saying what is not possible.

**Comment by h-vetinari** [comment]
The 2023 "state of the developer ecosystem" [report](https://www.jetbrains.com/lp/devecosystem-2023/cpp/) by JetBrains from a month ago underscores the usage patterns I was talking about. C++17 is by far the most widely used standard, and - more worringly - around 45% of users are not planning to migrate to newer standards from C++{11,14,17}. Even ignoring the incorrigible holdouts on C++{98,03}, that's quite a substantial chunk of the target market.

In part this reluctance is probably due to the (lack of) overall compiler maturity for C++20. Here's a recent [comment](https://discourse.llvm.org/t/rfc-clang-17-0-6-would-be-minimum-version-to-build-llvm-in-c-20/75345/8?u=h-vetinari) from someone at google involved in trying to upgrade their standard version to C++20, and how that's anything but a simple switch:
> There were multiple serious issues for consteval, concepts, and other new features. For Google’s internal codebase, these caused problems just trying to switch from -std=c++17 to -std=c++20 – including silent miscompiles. Our codebase didn’t even use C++20 features, we were just switching the compilation mode (which introduced usage of C++20 features via the stdlib headers.)
> 
> For reasonable C++20 compiler support, I’ve been making a recommendation of the following minimum compiler versions:
> 
>  * Clang 17
>  * GCC 12
>  * MSVC 17.2
>
> …and don’t use coroutines or modules yet, both of which are not yet fully-baked.

While the comment mostly refers to clang, it hopefully does illustrate the mismatch of the expectation that people will upgrade to C++20 and beyond quickly.

**Comment by jonwil** [comment]
Does the C++ standard say anything about whether valid C++17 code is also valid C++20 code?

**Comment by frederick-vs-ja** [comment]
> Does the C++ standard say anything about whether valid C++17 code is also valid C++20 code?

Roughly said in [[diff.cpp17]](https://eel.is/c++draft/diff.cpp17). (Looks off-topic here to me.)
